### PR TITLE
fix(cli): write-heading-id should not generate colliding slugs when not overwriting

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

Complete a TODO. This probably doesn't happen a lot, but better fix the edge cases

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

The test case is updated

### CODE REVIEW

File: packages/docusaurus/src/commands/writeHeadingIds.ts [hunk 1]
```diff
@@ -38,77 +38,60 @@ function addHeadingId(
 
   const headingText = line.slice(headingLevel).trimEnd();
   const headingHashes = line.slice(0, headingLevel);
-  const slug = slugger
-    .slug(unwrapMarkdownLinks(headingText).trim(), {maintainCase})
-    .replace(/^-+/, '')
```
Comment (Josh-Cena) [inline packages/docusaurus/src/commands/writeHeadingIds.ts:43]: This also has the possibility of generating colliding slugs, so it's removed. We would probably fix it inside the `slugger.slug` function by first filtering out the illegal characters

File: packages/docusaurus/src/commands/writeHeadingIds.ts [hunk 2]
```diff
@@ -38,77 +38,60 @@ function addHeadingId(
...
+  // If we can't overwrite existing slugs, make sure other headings don't
+  // generate colliding slugs by first marking these slugs as occupied
+  if (!overwrite) {
+    lines.forEach((line) => {
+      const parsedHeading = parseMarkdownHeadingId(line);
+      if (parsedHeading.id) {
+        slugger.slug(parsedHeading.id);
+      }
+    });
   }
```
Comment (Josh-Cena) [inline packages/docusaurus/src/commands/writeHeadingIds.ts:65]: Most of this is refactor and removing excess layers of abstraction, and this double-pass slugification is the only fix I made

File: packages/docusaurus/src/commands/__tests__/writeHeadingIds.test.ts [hunk 3]
```diff
@@ -113,14 +85,11 @@ describe('transformMarkdownContent', () => {
...
 
-    // TODO the first heading should probably rather be slugified to abc-1
-    // otherwise we end up with 2 x "abc" anchors
-    // not sure how to implement that atm
     const expected = `
 
 # Ignored title
 
-## abc {#abc}
+## abc {#abc-1}
```
Comment (slorber) [inline packages/docusaurus/src/commands/__tests__/writeHeadingIds.test.ts:92]: 👍 🥇

File: packages/docusaurus/src/commands/writeHeadingIds.ts (hunk 2 above)
Comment (slorber) [reply packages/docusaurus/src/commands/writeHeadingIds.ts:65]: 👍 nice trick

also wanted to use something similar unfortunately afaik the slugger does not offer an API to register forbidden slugs or something 😅

**Review by slorber — COMMENTED** [review]
Thanks 👍
